File: Leetcode/C++/125. Valid Palindrome/solution.cpp
// https://leetcode.com/problems/valid-palindrome/description/

#include <cctype>
#include <string>

using namespace std;

class Solution
{
public:
    bool isPalindrome(string s)
    {
        int left = 0, right = s.size() - 1;
        while (left < right)
        {
            // Move to alphanumeric characters
            while (left < right && !isalnum(s[left]))
            {
                left += 1;
            }
            while (left < right && !isalnum(s[right]))
            {
                right -= 1;
            }
            if (tolower(s[left]) != tolower(s[right])) // Check palindrome
            {
                return false;
            }
            left += 1;
            right -= 1;
        }
        return true;
    }
};

File: Leetcode/Python/125. Valid Palindrome/solution.py
# https://leetcode.com/problems/valid-palindrome/description/
# Time complexity: O(N)


import re


class Solution:  # Two pointer
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        left, right = 0, len(s) - 1
        while left < right:
            # Find index of alphanumeric characters
            while left < right and not s[left].isalnum():
                left += 1
            while right > left and not s[right].isalnum():
                right -= 1
            if s[left] != s[right]:  # Check palindrome
                return False
            left += 1
            right -= 1
        return True  # Checked all string return True


class Solution:  # Slicing
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        s = re.sub("[^a-z0-9]", "", s)
        return s == s[::-1]
